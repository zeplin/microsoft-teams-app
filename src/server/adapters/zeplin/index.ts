export * from "./zeplin";
export * from "./types";
