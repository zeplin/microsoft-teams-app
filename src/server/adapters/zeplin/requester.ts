import axios, { AxiosInstance, AxiosRequestConfig } from "axios";
import { ZeplinError } from "./ZeplinError";
import { INTERNAL_SERVER_ERROR } from "http-status-codes";

export class Requester {
    private instance: AxiosInstance;
    constructor(config: AxiosRequestConfig) {
        this.instance = axios.create(config);
    }

    async delete(url: string, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.delete(url, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async get<T = object>(url: string, config?: AxiosRequestConfig): Promise<T> {
        try {
            const { data } = await this.instance.get<T>(url, config);
            return data;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async patch(url: string, data: object, config?: AxiosRequestConfig): Promise<void> {
        try {
            await this.instance.patch(url, data, config);
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }

    async post(url: string, data: object, config?: AxiosRequestConfig): Promise<string> {
        try {
            const { data: { id } } = await this.instance.post<{ id: string }>(url, data, config);
            return id;
        } catch (error) {
            if (error.response) {
                throw new ZeplinError(error.response.data.message, { statusCode: error.response.status });
            } else {
                throw new ZeplinError(error.message, { statusCode: INTERNAL_SERVER_ERROR });
            }
        }
    }
}
