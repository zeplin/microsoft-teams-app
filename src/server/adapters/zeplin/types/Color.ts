export interface Color {
    id: string;
    created: number;
    name: string;
    r: number;
    g: number;
    b: number;
    a: number;
}
