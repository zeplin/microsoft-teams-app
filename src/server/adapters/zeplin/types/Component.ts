import { Thumbnails } from "./Thumbnails";

export interface Component {
    id: string;
    name: string;
    created: number;
    updated?: number;
    description?: string;
    image?: {
        width: number;
        height: number;
        original_url?: string;
        thumbnails?: Thumbnails;
    };
    section?: {
        id: string;
        group?: {
            id: string;
        };
    };
}
