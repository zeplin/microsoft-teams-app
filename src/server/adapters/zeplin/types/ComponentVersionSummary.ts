import { User } from "./User";
import { Thumbnails } from "./Thumbnails";

enum DestinationType {
    SCREEN = "screen",
    PREVIOUS = "previous",
}

export interface ComponentVersionSummary {
    id: string;
    created: number;
    creator?: User;
    commit?: {
        color: {
            r: number;
            g: number;
            b: number;
            a: number;
        };
        message?: string;
        author?: User;
    };
    image_url?: string;
    thumbnails?: Thumbnails;
    background_color?: {
        r: number;
        g: number;
        b: number;
        a: number;
    };
    source: string;
    width: number;
    height: number;
    density_scale: number;
    links: {
        rect: {
            x: number;
            y: number;
            width: number;
            height: number;
        };
        destination: {
            name?: string;
            type: DestinationType;
        };
    }[];
}
