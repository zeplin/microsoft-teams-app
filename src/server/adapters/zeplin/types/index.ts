export * from "./Organization";
export * from "./Project";
export * from "./Styleguide";
export * from "./RemPreferences";
export * from "./WebhookEvents";
export * from "./WebhookResourceType";
export * from "./ZeplinApplication";
export * from "./ProjectWebhook";
export * from "./StyleguideWebhook";
export * from "./User";
export * from "./WebhookStatus";
export * from "./WebhookUrlHealth";
