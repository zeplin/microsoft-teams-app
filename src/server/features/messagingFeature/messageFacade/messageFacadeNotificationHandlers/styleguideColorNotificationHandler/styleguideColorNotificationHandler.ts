import {
    ResourceType,
    WebhookEvent,
    EventPayload,
    StyleguideContext,
    EventType
} from "../../../messagingTypes";
import { NotificationHandler } from "../NotificationHandler";
import { SHORT_DELAY } from "../constants";
import { commonTeamsCard, AdaptiveCard } from "../teamsCardTemplates";
import { ZEPLIN_WEB_APP_BASE_URL, ZEPLIN_MAC_APP_URL_SCHEME } from "../../../../../config";
import { URL } from "url";

type StyleguideColorEventDescriptor = {
    type: EventType.STYLEGUIDE_COLOR;
    action: "created" | "updated";
};

type StyleguideColorResource = {
    id: string;
    type: ResourceType.COLOR;
    data: {
        id: string;
        created: number;
        name: string;
        r: number;
        g: number;
        b: number;
        a: number;
    };
};

class StyleguideColorNotificationHandler extends NotificationHandler {
    delay = SHORT_DELAY;
    private getText(events: WebhookEvent<StyleguideColorEventPayload>[]): string {
        const [{
            payload: {
                action,
                context: {
                    styleguide: {
                        name: styleguideName
                    }
                },
                resource: {
                    data: {
                        name: pivotColorName
                    }
                }
            }
        }] = events;
        const actionText = action === "created" ? "added" : "updated";
        return events.length === 1
            ? `**${pivotColorName}** is ${actionText} in _${styleguideName}_! üèÉ‚Äç‚ôÇ`
            : `**${events.length} new colors** are ${actionText} in _${styleguideName}_! üèÉ‚Äç‚ôÇ`;
    }

    private getWebappURL(
        events: WebhookEvent<StyleguideColorEventPayload>[]
    ): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        id: styleguideId
                    }
                }
            }
        }] = events;
        const webappURL = new URL(ZEPLIN_WEB_APP_BASE_URL);
        webappURL.pathname = `styleguide/${styleguideId}/styleguide/colors`;
        events.forEach(event => webappURL.searchParams.append("cid", event.payload.resource.id));
        return webappURL.toString();
    }

    private getMacAppURL(
        events: WebhookEvent<StyleguideColorEventPayload>[]
    ): string {
        const [{
            payload: {
                context: {
                    styleguide: {
                        id: styleguideId
                    }
                }
            }
        }] = events;
        return `${ZEPLIN_MAC_APP_URL_SCHEME}colors?stid=${styleguideId}&cids=${events.map(event => event.payload.resource.id).join(",")}`;
    }

    shouldHandleEvent(event: WebhookEvent): boolean {
        return event.payload.action !== "deleted";
    }

    getTeamsMessage(
        events: WebhookEvent<StyleguideColorEventPayload>[]
    ): AdaptiveCard {
        return commonTeamsCard({
            text: this.getText(events),
            sectionText: "Make sure your stylesheets are up to date!",
            links: [{
                title: "Open in App",
                url: this.getMacAppURL(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }
}

export type StyleguideColorEventPayload = EventPayload<
    StyleguideColorEventDescriptor,
    StyleguideContext,
    StyleguideColorResource
>;
export const styleguideColorNotificationHandler = new StyleguideColorNotificationHandler();