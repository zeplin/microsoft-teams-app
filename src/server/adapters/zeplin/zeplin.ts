import { ProjectWebhooks } from "./ProjectWebhooks";
import { Requester } from "./requester";
import { StyleguideWebhooks } from "./StyleguideWebhooks";

interface ZeplinOptions {
    url: string;
    webhookSecret: string;
}

class Zeplin {
    public projectWebhooks!: ProjectWebhooks;
    public styleguideWebhooks!: StyleguideWebhooks;

    init({ url, webhookSecret }: ZeplinOptions): void {
        const requester = new Requester({ baseURL: `${url}/v1/` });
        this.projectWebhooks = new ProjectWebhooks(requester, webhookSecret);
        this.styleguideWebhooks = new StyleguideWebhooks(requester, webhookSecret);
    }
}

export const zeplin = new Zeplin();
