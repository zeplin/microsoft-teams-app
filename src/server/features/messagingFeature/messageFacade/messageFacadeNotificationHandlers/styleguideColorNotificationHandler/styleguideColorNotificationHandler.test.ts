import { styleguideColorNotificationHandler } from "./styleguideColorNotificationHandler";
import { StyleguideColorEvent } from "../../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    action?: string;
    colorId?: string;
    colorName?: string;
}

function getDummyEvent({
    action = "created",
    colorId = "colorId",
    colorName = "colorName"
}: GetDummyEventParams = {}): StyleguideColorEvent {
    return {
        payload: {
            action,
            context: {
                styleguide: {
                    id: "styleguideId",
                    name: "styleguideName"
                }
            },
            resource: {
                id: colorId,
                data: {
                    id: colorId,
                    name: colorName
                }
            }
        }
    } as StyleguideColorEvent;
}

describe("styleguideColorNotificationHandler", () => {
    describe("getTeamsMessage method", () => {
        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    styleguideColorNotificationHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    styleguideColorNotificationHandler.getTeamsMessage([
                        getDummyEvent(),
                        getDummyEvent({
                            colorId: "anotherColorId",
                            colorName: "anotherColorName"
                        })
                    ])
                ).toMatchSnapshot();
            });
        });

        describe("for updated notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    styleguideColorNotificationHandler.getTeamsMessage([getDummyEvent({ action: "updated" })])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    styleguideColorNotificationHandler.getTeamsMessage([
                        getDummyEvent({
                            action: "updated"
                        }),
                        getDummyEvent({
                            action: "updated",
                            colorId: "anotherColorId",
                            colorName: "anotherColorName"
                        })
                    ])
                ).toMatchSnapshot();
            });
        });
    });
});
