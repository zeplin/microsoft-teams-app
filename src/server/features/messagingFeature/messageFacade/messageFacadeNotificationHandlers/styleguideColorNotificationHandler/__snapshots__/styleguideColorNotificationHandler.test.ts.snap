// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styleguideColorNotificationHandler getTeamsMessage method for created notification should match snapshot when there are more than 1 events 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**2 new colors** are added in _styleguideName_! üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Make sure your stylesheets are up to date!",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://colors?stid=styleguideId&cids=colorId,anotherColorId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/styleguide/styleguideId/styleguide/colors?cid=colorId&cid=anotherColorId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorNotificationHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**colorName** is added in _styleguideName_! üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Make sure your stylesheets are up to date!",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://colors?stid=styleguideId&cids=colorId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/styleguide/styleguideId/styleguide/colors?cid=colorId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorNotificationHandler getTeamsMessage method for updated notification should match snapshot when there are more than 1 events 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**2 new colors** are updated in _styleguideName_! üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Make sure your stylesheets are up to date!",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://colors?stid=styleguideId&cids=colorId,anotherColorId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/styleguide/styleguideId/styleguide/colors?cid=colorId&cid=anotherColorId"
        }
      ]
    }
  ]
}
`;

exports[`styleguideColorNotificationHandler getTeamsMessage method for updated notification should match snapshot when there is only 1 event 1`] = `
{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "**colorName** is updated in _styleguideName_! üèÉ‚Äç‚ôÇ",
      "wrap": true,
      "spacing": "small"
    },
    {
      "type": "Container",
      "spacing": "extraLarge",
      "items": [
        {
          "type": "TextBlock",
          "text": "Make sure your stylesheets are up to date!",
          "spacing": "none"
        }
      ]
    },
    {
      "type": "ActionSet",
      "spacing": "extraLarge",
      "actions": [
        {
          "type": "Action.OpenUrl",
          "title": "Open in App",
          "url": "zpl://colors?stid=styleguideId&cids=colorId"
        },
        {
          "type": "Action.OpenUrl",
          "title": "Open in Web",
          "url": "https://app.zeplin.io/styleguide/styleguideId/styleguide/colors?cid=colorId"
        }
      ]
    }
  ]
}
`;
